8394 maven mods

poikilotherm opened the pull request:
**What this PR does / why we need it**:
This PR will introduce a new Maven parent POM and refactor the current main POM to use it. Basically nothing really changes from this refactoring despite the fact that we may rely on this new module to coordinate dependency versions, dependency convergence, importing common BOMs etc.

This PR also adds a filter to the Github Action to run unit tests when the main Dataverse POM and the POMs in the modules are changed. This might be even changed to make the unit tests run on the parent POM, so it picks up all modules. (@donsizemore WDYT?)

Please note: this PR maybe should incorporate changes to include the POM for ZIP downloader as a submodule. As it might make sense to create another module in `module/dataverse-zipdownloader` or similar, this might also live within another pull request. Small steps, wholesome stuff - whatever you prefer.

**Which issue(s) this PR closes**:

Relates to #8394 

(Not using "Closes" here because of the remaining things like refactoring ZIP downloader)

**Special notes for your reviewer**:
None so far. Please add any suggestions, opinions etc. Should we add some doccs in the dev guide -> dependency management? What about adding some hints on architecture, as this might be a start to move Dataverse monolith towards a modulith?

**Suggestions on how to test this**:
Use Maven in project root, watch it do as it always has. Go to `modules/dataverse-parent` and use Maven there, see the difference (submodules detected and used).

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:
Nope.

**Is there a release notes update needed for this change?**:
Not yet. Please add suggestions on this matter...

**Additional documentation**:
None yet.

poikilotherm commented:
I have no idea why this won't build on Jenkins - cannot look in the logs, either. Sorry. Unit tests are doing OK

poikilotherm commented:
Hey @landreev I just took to opportunity to refactor the zipper submodule with the parent to extend this example a bit further, outlining further benefits.

In the same go slightly refactored the packaging to use Spring Boot Fat JAR generation instead of Maven Assembly.
If you like this refactoring, we maybe should change the guides a bit?

(And maybe talk about releasing this component? As it's in an experimental status, it might be worth changing the version to sth <1.0? )

landreev commented:
Apologies for the delay. I wanted to build the zipper (it builds for me w/out problems, yes) and test it a bit (still need to do some more of that). The refactoring generally seems ok. I'm not sure about the name change; I do like your "zipdownloader" better - but seeing how we mentioned it under the old name, and it would need to be changed - I'm not sure it's worth it. 
(How it ended up with "ZipDownloadService" as the group id - that I don't even know). 

As for releasing it... in all honesty, I'm very apprehensive about advertising it to other installations as an integral part of the application. In its current form it's a big hack; (by design, but a hack nevertheless). If we wanted to release it and make it an official, recommended application component - and not just something a very adventurous installation can choose to use at their own risk, I would revisit certain solutions there first. (We can certainly talk about it some more outside of this issue). 
(The fact that it runs under cgi-bin though I don't see a problem with, TBH; it was part of the minimalist/super simplified/zero overhead approach behind the whole thing.)

landreev commented:
As for the rest of the PR, I'm still seeing things like `slf4j-log4j12` in the new `dataverse-parent/pom.xml`. I'm assuming that needs to be updated, since #8377 was merged - ?

poikilotherm commented:
@landreev thanks for taking a look, much appreciated!

> Apologies for the delay. I wanted to build the zipper (it builds for me w/out problems, yes) and test it a bit (still need to do some more of that). The refactoring generally seems ok. 

:smile: 

> I'm not sure about the name change; I do like your "zipdownloader" better - but seeing how we mentioned it under the old name, and it would need to be changed - I'm not sure it's worth it. (How it ended up with "ZipDownloadService" as the group id - that I don't even know).
> 
> As for releasing it... in all honesty, I'm very apprehensive about advertising it to other installations as an integral part of the application. In its current form it's a big hack; (by design, but a hack nevertheless). If we wanted to release it and make it an official, recommended application component - and not just something a very adventurous installation can choose to use at their own risk, I would revisit certain solutions there first. (We can certainly talk about it some more outside of this issue).
(The fact that it runs under cgi-bin though I don't see a problem with, TBH; it was part of the minimalist/super simplified/zero overhead approach behind the whole thing.)

I still really like the idea - having a separated component handling the downloads is great. Yes, it's hacky, but it outlines the idea and is a start. As you mentioned: it would be better discussed in a separate issue.

For the name change: I think it IS worth the change. The change is minimal and I need to update the docs anyway (introduce modules and parent within the dev guide). Now is the perfect chance to switch the artifact id and group id, in preparation to craft a non-hacky solution in the future.

> As for the rest of the PR, I'm still seeing things like slf4j-log4j12 in the new dataverse-parent/pom.xml. I'm assuming that needs to be updated, since https://github.com/IQSS/dataverse/pull/8377 was merged - ?

I re-incorporated the changes, goot catch, thanks! Not sure why they didn't trigger a merge conflict... Anyway, I rebased on `develop` and force-pushed to make sure this works smoothly when merged without any conflicts. (You'll need to reset hard/rebase to the PR branch if you want to pull the changes)

landreev commented:
OK, zipper is still working as an executable/under cgi-bin, when compiled with springframework boot... I mean, there was no reason to expect otherwise, but I wanted to confirm.

landreev commented:
> I still really like the idea - having a separated component handling the downloads is great. 

The backstory behind this thing is that they (here, at IQSS) wanted to give users a one-click way to download **gigantic** datasets, as zip files.  I honestly thought it was a bad idea. But as a compromise I put together this standalone tool that would at least offload this zipping and streaming task from the application server(s).  

> Yes, it's hacky, but it outlines the idea and is a start. 

We've been talking (for a while) about making the application modular/making it possible to run certain individual components on separate servers. Having a dedicated server running the download API would be a good example. That would be great to have and we should continue discussing/working on it. But just to be clear, this particular tool is NOT a start of that modularization effort. This thing is not a Dataverse component - it doesn't know, and doesn't need to know, anything about internal Dataverse objects, permissions or its APIs, etc. It's a very narrow-purpose, one-trick standalone tool that simply allows us to handle multi-file zipped downloads the same way we handle individual "direct downloads" of files stored on S3. I.e. it allows Dataverse to issue a redirect and forget about it. It's been useful in that capacity, both to us and a couple of other installations, so let's keep and maintain it as such. But we should continue talking about creating "true" application components in parallel, separately.

poikilotherm commented:
Today I added a really large section within the docs about all this:

![grafik](https://user-images.githubusercontent.com/5468333/153455889-bf8f17cf-eeb8-453d-82d5-f9e29a5df9f8.png)

landreev commented:
I really like this Dependency Management guide entry.

poikilotherm commented:
@landreev I'm sorry it looks like I accidentially nuked your commit... Sorry!!!

Fixed the script again and refactored the docs about zipdownloader.

landreev commented on doc/sphinx-guides/source/installation/advanced.rst:
```diff
@@ -68,43 +68,45 @@ Once a standardized version of you Custom Terms are registered as a license, an
+
+1. Follow the instructions in the file above to build ``zipdownloader-0.0.1.jar``. (Also available from
+   `zipper.zip <https://github.com/IQSS/dataverse/releases/download/v5.0/zipper.zip>`_ of the
+   `Dataverse Software 5.0 release on GitHub <https://github.com/IQSS/dataverse/releases/tag/v5.0>`_).
```
I would prefer not to modify this file (zipper.zip) that we released as part of 5.0. In part because we want this to work as advertised in the 5.0-5.9 versions of the guide. 
Let me think of how to modify this section. I'm thinking of mentioning that we have renamed this jar file as of this release; and that it used to be distributed under a different name; but everything should still be working the same...

poikilotherm replied:
FWIW: this hasn't been part of any releases since 5.0, which is why I linked to it directly.

landreev replied:
I made a couple of small changes to this section.

pdurbin commented:
I took a look a the docs changes and made a pull request with some suggestions: https://github.com/poikilotherm/dataverse/pull/524

pdurbin commented:
@poikilotherm and I iterated a bit on https://github.com/poikilotherm/dataverse/pull/524 and I hope it's ready to merge.

Meanwhile, why is this pull request still in draft? Are we waiting for something?

At standup yesterday I believe we said we're waiting to hear a bit more about the vision for this change. I'm not sure what form this would take. A discussion at tech hours? More docs? A comment here?

Finally, I'll just add that I was able to build and deploy this branch in docker-aio. So nothing seems to be broken, from a quick check. I appreciate that the parent pom is on the side and that we don't have to immediately change how we do things, where we invoke `mvn package`, etc.

I guess I'll leave myself on this pull request for now but I don't have any more to say. I think it's ok to merge (but again, I'd like https://github.com/poikilotherm/dataverse/pull/524 to be merged first).

poikilotherm commented:
@pdurbin this is still in draft as there might be things left you folks don't like.

And maybe let's talk about this stuff first before creating facts and revokes if consensus says we don't wanna go down this road.

pdurbin commented:
@poikilotherm sounds good. From standup, it sounds like we're looking forward to you talking a bit about this pull request at a future tech hours before we merge. I'll assign this to you for now.

Thanks for merging my doc change. I don't think I have anything else to add so I'll unassign myself.

@landreev mentioned he had a least a doc change to make.

pdurbin commented:
@poikilotherm are you ready to move this PR out of Draft status? Yesterday at standup @landreev and I said we think it can go to QA.

pdurbin commented:
@poikilotherm heads up that in 09167d5 I reverted a Jacoco-related commit that may be the reason the Jenkins job is failing. I left details in the comment.

pdurbin reviewed (approved):
The Jenkins tests are passing now so I'm approving this. As discussed at standup yesterday @landreev seemed ok with it too.

poikilotherm commented:
Oh hey great thanks for merging!

Pretty cool you folks let me introduce this change 😊

I promise to be helpful when question arise in the future 🙈✌️
